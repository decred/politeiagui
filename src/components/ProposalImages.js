import React from "react";

const ProposalImages = ({ files, ...props }) => (
  <div {...props}>
    {(files || []).map(({ name, mime, digest, payload }, idx) => (
      <div key={digest || idx}>
        <h5>{name}</h5>
        <img alt={name} src={`data:${mime};base64,${payload}`} />
      </div>
    ))}
  </div>
);

export default ProposalImages;
