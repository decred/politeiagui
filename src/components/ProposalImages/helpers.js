/* The validation will return an object with the following shape
  { files: [file1, file2], validationErrors: ['message1', 'message2']}
*/
export const errorTypes = {
  MAX_SIZE: 'max_size',
  MAX_IMAGES: 'max_length'
}

export function validateFiles(files, policy) {
  const validation = {
    files,
    errors: []
  }
  const validatedMaxImages = validateMaxImages(validation, policy);
  const validatedMaxSize = validateMaxSize(validatedMaxImages, policy);
  return validatedMaxSize;
}

export function getFormattedFiles({ base64, fileList }) {
  return Array.from(fileList).map(({ name, size, type: mime }, idx) => ({
    name, mime, size, payload: base64[idx].split("base64,").pop()
  }))
}

export function isFileValid(file, policy) {
  if (file.size > policy.maximagesize) {
    return false;
  }

  if (policy.validmimetypes.indexOf(file.mime) < 0) {
    return false;
  }

  return true;
}

function getErrorMessage(policy, errorType, filename = '') {
  const errors = {
    [errorTypes.MAX_SIZE]: `The file ${filename} exceeds the maximum size`,
    [errorTypes.MAX_IMAGES]: `You can upload a maximum of ${policy.maximages} images per proposal`
  }
  return errors[errorType];
}

function validateMaxImages({errors, files}, policy) {
  if (files.length > policy.maximages) {
    errors.push(getErrorMessage(policy, errorTypes.MAX_IMAGES));
    return ({
      files: files.slice(0, policy.maximages - 1),
      errors
    });
  }
  return ({errors, files});
}

function validateMaxSize({errors, files}, policy) {
  const newFiles = files.filter(file => {
    if (file.size > policy.maximagesize) {
      errors.push(getErrorMessage(policy, errorTypes.MAX_SIZE, file.name));
      return false;
    }
    return true;
  })
  return ({
    files: newFiles,
    errors
  });
}
