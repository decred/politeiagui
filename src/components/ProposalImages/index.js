import React, { Component } from "react";
import PropTypes from "prop-types";
import { isFileValid } from "./helpers";

class ProposalImages extends Component {
  constructor(props) {
    super(props);

    this.onRemove = this.onRemove.bind(this);
  }

  onRemove(idx) {
    this.props.onChange({ remove: idx });
  }

  render() {
    const { files, readOnly, policy } = this.props;

    return (
      <div>
        {(files || []).map(({ name, mime, digest, payload, size }, idx) => (
          <div key={digest || idx}>
            <h5>{name}{readOnly ? null : isFileValid({ size, mime }, policy) ? null :  <span className="error">&nbsp;Errored</span> }</h5>
            <div className="attached-image-ct clearfloat">
              <img className="attached-image" alt={name} src={`data:${mime};base64,${payload}`} />
              {readOnly ? null : (
                <a className="attached-image-remove" onClick={() => this.onRemove(idx)} title="Remove image">âœ–</a>
              )}
            </div>
          </div>
        ))}
      </div>
    );
  }
}

ProposalImages.propTypes = {
  files: PropTypes.array.isRequired,
  policy: PropTypes.object,
  readOnly: PropTypes.bool.isRequired,
};

ProposalImages.defaultProps = {
  readOnly: false,
  files: [],
};

export default ProposalImages;
