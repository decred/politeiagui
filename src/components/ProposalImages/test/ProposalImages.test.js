import React from "react";
import ProposalImages from "../index";
import renderer from "react-test-renderer";

const maliciousSVGFile = {
  name: "malicious",
  mime: "image/svg+xml",
  payload: `PD94bWwgdmVyc2lvbj0iMS4wIiBzdGFuZGFsb25lPSJubyI/Pgo8IURPQ1RZUEU
  gc3ZnIFBVQkxJQyAiLS8vVzNDLy9EVEQgU1ZHIDEuMS8vRU4iICJodHRwOi8vd3d3LnczLm9y
  Zy9HcmFwaGljcy9TVkcvMS4xL0RURC9zdmcxMS5kdGQiPgoKPHN2ZyB2ZXJzaW9uPSIxLjEiI
  GJhc2VQcm9maWxlPSJmdWxsIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPg
  ogICA8cG9seWdvbiBpZD0idHJpYW5nbGUiIHBvaW50cz0iMCwwIDAsNTAgNTAsMCIgZmlsbD0
  iIzAwOTkwMCIgc3Ryb2tlPSIjMDA0NDAwIi8+CiAgIDxzY3JpcHQgdHlwZT0idGV4dC9qYXZh
  c2NyaXB0Ij4KICAgIHdpbmRvdy5sb2NhdGlvbi5yZXBsYWNlKCJodHRwOi8vd3d3Lmdvb2dsZ
  S5jb20iKTs7CiAgIDwvc2NyaXB0Pgo8L3N2Zz4K`,
  size: 405
};

const notMaliciousFile = {
  name: "notMalicious",
  mime: "image/svg+xml",
  payload: `PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPgogICA8cG9
  seWdvbiBzdHJva2U9IiMwMDQ0MDAiIGZpbGw9IiMwMDk5MDAiIHBvaW50cz0iMCwwIDAsNTAg
  NTAsMCIgaWQ9InRyaWFuZ2xlIj48L3BvbHlnb24+CiAgIAo8L3N2Zz4K`
};

it("sanitize SVG files", () => {
  const tree = renderer
    .create(
      <ProposalImages
        files={[maliciousSVGFile, notMaliciousFile]}
        onChange={() => null}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
