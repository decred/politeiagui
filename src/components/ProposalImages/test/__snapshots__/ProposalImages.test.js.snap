// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanitize SVG files 1`] = `
<div
  className="attached-images"
>
  <div
    className="attached-image-ct"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <h5
        className="attached-image-title"
      >
        malicious
      </h5>
      <a
        className="attached-image-remove"
        onClick={[Function]}
        title="Remove image"
      >
        ✖
      </a>
    </div>
    <img
      alt="malicious"
      className="attached-image"
      src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPgogICA8cG9seWdvbiBzdHJva2U9IiMwMDQ0MDAiIGZpbGw9IiMwMDk5MDAiIHBvaW50cz0iMCwwIDAsNTAgNTAsMCIgaWQ9InRyaWFuZ2xlIj48L3BvbHlnb24+CiAgIAo8L3N2Zz4K"
    />
  </div>
  <div
    className="attached-image-ct"
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
        }
      }
    >
      <h5
        className="attached-image-title"
      >
        notMalicious
      </h5>
      <a
        className="attached-image-remove"
        onClick={[Function]}
        title="Remove image"
      >
        ✖
      </a>
    </div>
    <img
      alt="notMalicious"
      className="attached-image"
      src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPgogICA8cG9seWdvbiBpZD0idHJpYW5nbGUiIHBvaW50cz0iMCwwIDAsNTAgNTAsMCIgZmlsbD0iIzAwOTkwMCIgc3Ryb2tlPSIjMDA0NDAwIj48L3BvbHlnb24+CiAgIAo8L3N2Zz4K"
    />
  </div>
</div>
`;
